File: android/core/src/main/kotlin/dev/yikz/clipboard/core/SyncEngine.kt
package dev.yikz.clipboard.core

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.random.Random

interface SyncHandler {
    suspend fun onWelcome(welcome: WelcomeMsg)
    suspend fun onMessage(message: WsMessage)
    suspend fun onDisconnected()
}

interface PreparedClip {
    suspend fun write()
}

interface ItemApplier {
    fun autoDownloadLimit(): Long
    fun isKindEnabled(kind: String): Boolean
    suspend fun prepare(item: CachedItem): PreparedClip?
    fun offerDownload(item: CachedItem)
}

interface SyncCallbacks {
    fun onItemsChanged() {}
    fun onUnauthorized() {}
    fun onKeyInvalid() {}
    fun onStorageWarning(message: StorageWarningMsg) {}
    fun onDevicesChanged() {}
    fun onApplied(item: CachedItem) {}
    fun onReleaseAvailable(version: String) {}
}

class SyncEngine(
    private val api: SyncApi,
    private val store: SyncStore,
    private val codec: ItemCodec,
    private val ownDeviceId: String,
    private val saltB64: String,
    private val applier: ItemApplier,
    private val echo: EchoGuard,
    private val scope: CoroutineScope,
    private val clock: () -> Long = System::currentTimeMillis,
    private val log: Logger = NoLog,
    private val callbacks: SyncCallbacks = object : SyncCallbacks {},
    private val initialHistoryTarget: Int = 1000,
    private val random: Random = Random.Default,
) : SyncHandler {
    private val mutex = Mutex()
    private val reconcileMutex = Mutex()
    private var catchingUp = false
    private val liveDuringCatchUp = ArrayList<CachedItem>()
    private var catchUpJob: Job? = null
    private var highestWrittenSeq = 0L
    private var reconciling = false
    private val revsDuringReconcile = HashSet<Long>()
    private val pinsDuringReconcile = ArrayList<ClipPinnedMsg>()
    private val _syncing = MutableStateFlow(false)

    @Volatile
    var clockOffsetMs: Long = 0
        private set

    val syncing: StateFlow<Boolean> = _syncing.asStateFlow()

    fun serverNow(): Long = clock() + clockOffsetMs

    override suspend fun onWelcome(welcome: WelcomeMsg) {
        catchUpJob?.cancelAndJoin()
        mutex.withLock {
            Timestamps.parseOrNull(welcome.serverTime)?.let { clockOffsetMs = it - clock() }
            catchingUp = true
            liveDuringCatchUp.clear()
        }
        _syncing.value = true
        catchUpJob = scope.launch { catchUpWithRetry(welcome) }
    }

    override suspend fun onDisconnected() {
        catchUpJob?.cancelAndJoin()
        catchUpJob = null
        mutex.withLock {
            catchingUp = false
            liveDuringCatchUp.clear()
        }
        _syncing.value = false
    }

    override suspend fun onMessage(message: WsMessage) {
        when (message) {
            is ClipMsg -> onClip(message.item)
            is ClipDeletedMsg -> onDeleted(message)
            is ClipPinnedMsg -> onPinned(message)
            is StorageWarningMsg -> callbacks.onStorageWarning(message)
            is DevicesChangedMsg -> callbacks.onDevicesChanged()
            is ReleaseAvailableMsg -> callbacks.onReleaseAvailable(message.version)
            else -> Unit
        }
    }

    suspend fun awaitCatchUp() {
        catchUpJob?.join()
    }

    private suspend fun catchUpWithRetry(welcome: WelcomeMsg) {
        var attempt = 0
        while (true) {
            try {
                runCatchUp(welcome)
                return
            } catch (e: CancellationException) {
                throw e
            } catch (e: ApiException) {
                if (e.isUnauthorized) {
                    callbacks.onUnauthorized()
                    return
                }
                log.w(TAG, "catch-up failed: ${e.code} ${e.message}")
            } catch (e: StopSync) {
                return
            } catch (e: Exception) {
                log.e(TAG, "catch-up failed", e)
            }
            delay(Backoff.httpRetryDelayMs(attempt++, random))
        }
    }

    private class StopSync : Exception()

    private suspend fun runCatchUp(welcome: WelcomeMsg) {
        var state = mutex.withLock { store.state() }
        if (state.serverId != null && state.serverId != welcome.serverId) {
            log.w(TAG, "server id changed, clearing local cache")
            mutex.withLock {
                store.clearAll()
                state = SyncState(serverId = welcome.serverId)
                store.saveState(state)
                highestWrittenSeq = 0
            }
            callbacks.onItemsChanged()
            if (!verifyKeyAgainstServer()) {
                callbacks.onKeyInvalid()
                throw StopSync()
            }
        } else if (state.serverId == null) {
            mutex.withLock {
                state = store.state().copy(serverId = welcome.serverId)
                store.saveState(state)
            }
        }

        val initial = state.lastSeq == 0L
        val inserted = ArrayList<CachedItem>()
        var cursor = state.lastSeq
        if (initial) {
            var before = welcome.currentSeq + 1
            var count = 0
            while (welcome.currentSeq > 0) {
                val page = api.history(before = before, limit = Protocol.HISTORY_DEFAULT_LIMIT)
                val items = page.items.map(codec::decode)
                mutex.withLock { store.upsert(items) }
                items.forEach { echo.add(it.contentHash) }
                callbacks.onItemsChanged()
                count += items.size
                if (!page.hasMore || items.isEmpty() || count >= initialHistoryTarget) break
                before = items.minOf { it.seq }
            }
            cursor = welcome.currentSeq
            log.i(TAG, "initial sync loaded $count items")
        } else if (welcome.currentSeq > cursor) {
            while (true) {
                val page = api.history(after = cursor, limit = Protocol.HISTORY_MAX_LIMIT)
                val items = page.items.map(codec::decode)
                mutex.withLock { store.upsert(items) }
                items.forEach { echo.add(it.contentHash) }
                inserted += items
                if (items.isNotEmpty()) cursor = items.maxOf { it.seq }
                callbacks.onItemsChanged()
                if (!page.hasMore || items.isEmpty()) break
            }
            log.i(TAG, "catch-up loaded ${inserted.size} items after seq ${state.lastSeq}")
        }

        if (welcome.stateRev != state.stateRev) reconcile()

        val pick = mutex.withLock {
            val current = store.state()
            val live = liveDuringCatchUp.toList()
            val candidates = if (initial) live else inserted + live
            val maxSeen = maxOf(candidates.maxOfOrNull { it.seq } ?: 0L, if (initial) welcome.currentSeq else 0L)
            val chosen = AutoApply.pickAfterCatchUp(candidates, ownDeviceId, current.appliedSeq, serverNow())
            store.saveState(
                current.copy(
                    lastSeq = maxOf(current.lastSeq, cursor, live.maxOfOrNull { it.seq } ?: 0L),
                    appliedSeq = maxOf(current.appliedSeq, maxSeen),
                ),
            )
            catchingUp = false
            liveDuringCatchUp.clear()
            chosen
        }
        _syncing.value = false
        pick?.let { launchApply(it) }
    }

    private suspend fun verifyKeyAgainstServer(): Boolean {
        val me = api.me()
        if (me.salt != saltB64) return false
        val serverCheck = me.keyCheck
        if (serverCheck == null) {
            try {
                api.putKeyCheck(codec.key.keyCheck)
                return true
            } catch (e: ApiException) {
                if (e.code != "key_check_exists") throw e
                return api.me().keyCheck == codec.key.keyCheck
            }
        }
        return serverCheck == codec.key.keyCheck
    }

    suspend fun reconcile() {
        reconcileMutex.withLock {
            mutex.withLock {
                reconciling = true
                revsDuringReconcile.clear()
                pinsDuringReconcile.clear()
            }
            try {
                val index = api.historyIndex()
                mutex.withLock {
                    store.reconcile(index.items.associate { it.id to it.pinned }, index.currentSeq)
                    for (pin in pinsDuringReconcile) {
                        if (pin.stateRev > index.stateRev) store.setPinned(pin.id, pin.pinned)
                    }
                    var rev = index.stateRev
                    while (rev + 1 in revsDuringReconcile) rev++
                    store.saveState(store.state().copy(stateRev = rev))
                }
                log.i(TAG, "reconciled with state_rev ${index.stateRev}, ${index.items.size} items on server")
                callbacks.onItemsChanged()
            } finally {
                mutex.withLock {
                    reconciling = false
                    revsDuringReconcile.clear()
                    pinsDuringReconcile.clear()
                }
            }
        }
    }

    private suspend fun onClip(header: ItemHeader) {
        val item = codec.decode(header)
        if (item.metaError != null) log.e(TAG, "cannot decrypt meta of ${item.id}: ${item.metaError}")
        echo.add(item.contentHash)
        val toApply = mutex.withLock {
            store.upsert(listOf(item))
            if (catchingUp) {
                liveDuringCatchUp += item
                null
            } else {
                val state = store.state()
                val eligible = AutoApply.isEligible(item, ownDeviceId, state.appliedSeq, serverNow())
                store.saveState(
                    state.copy(
                        lastSeq = maxOf(state.lastSeq, item.seq),
                        appliedSeq = maxOf(state.appliedSeq, item.seq),
                    ),
                )
                if (eligible) item else null
            }
        }
        callbacks.onItemsChanged()
        toApply?.let { launchApply(it) }
    }

    private suspend fun onDeleted(message: ClipDeletedMsg) {
        val needReconcile = mutex.withLock {
            store.delete(message.ids)
            if (reconciling) revsDuringReconcile += message.stateRev
            handleRev(message.stateRev)
        }
        callbacks.onItemsChanged()
        if (needReconcile) launchReconcile()
    }

    private suspend fun onPinned(message: ClipPinnedMsg) {
        val needReconcile = mutex.withLock {
            store.setPinned(message.id, message.pinned)
            if (reconciling) {
                revsDuringReconcile += message.stateRev
                pinsDuringReconcile += message
            }
            handleRev(message.stateRev)
        }
        callbacks.onItemsChanged()
        if (needReconcile) launchReconcile()
    }

    private fun handleRev(rev: Long): Boolean {
        val state = store.state()
        val current = state.stateRev
        return when {
            current != null && rev == current + 1 -> {
                store.saveState(state.copy(stateRev = rev))
                false
            }
            current != null && rev <= current -> false
            else -> !catchingUp && !reconciling
        }
    }

    private fun launchReconcile() {
        scope.launch {
            try {
                reconcile()
            } catch (e: CancellationException) {
                throw e
            } catch (e: ApiException) {
                if (e.isUnauthorized) callbacks.onUnauthorized() else log.w(TAG, "reconcile failed: ${e.code}")
            } catch (e: Exception) {
                log.e(TAG, "reconcile failed", e)
            }
        }
    }

    private fun launchApply(item: CachedItem) {
        if (item.meta == null) return
        if (!applier.isKindEnabled(item.kind)) {
            log.i(TAG, "not applying ${item.id}: ${item.kind} sync is off")
            return
        }
        when (AutoApply.action(item, applier.autoDownloadLimit())) {
            ApplyAction.OFFER_DOWNLOAD -> applier.offerDownload(item)
            else -> scope.launch { apply(item, guard = true) }
        }
    }

    suspend fun applyNow(item: CachedItem): Boolean = apply(item, guard = false)

    private suspend fun apply(item: CachedItem, guard: Boolean): Boolean {
        val prepared = try {
            applier.prepare(item)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            log.e(TAG, "could not prepare ${item.id}", e)
            null
        } ?: return false
        val allowed = mutex.withLock {
            if (guard && item.seq < highestWrittenSeq) {
                false
            } else {
                highestWrittenSeq = maxOf(highestWrittenSeq, item.seq)
                true
            }
        }
        if (!allowed) {
            log.i(TAG, "skipping ${item.id}: a newer item was applied meanwhile")
            return false
        }
        echo.add(item.contentHash)
        prepared.write()
        callbacks.onApplied(item)
        return true
    }

    suspend fun loadOlder(): Boolean {
        val oldest = mutex.withLock { store.oldestSeq() } ?: return false
        if (oldest <= 1) return false
        val page = api.history(before = oldest, limit = Protocol.HISTORY_DEFAULT_LIMIT)
        val items = page.items.map(codec::decode)
        mutex.withLock { store.upsert(items) }
        callbacks.onItemsChanged()
        return page.hasMore && items.isNotEmpty()
    }

    companion object {
        private const val TAG = "sync"
    }
}

File: windows/src/YikzClipboard.Core/Sync/SyncEngine.cs
using System.Text.Json;
using YikzClipboard.Core.Crypto;
using YikzClipboard.Core.Logging;
using YikzClipboard.Core.Net;
using YikzClipboard.Core.Platform;
using YikzClipboard.Core.Protocol;
using YikzClipboard.Core.Storage;

namespace YikzClipboard.Core.Sync;

public sealed class SyncContext
{
    public required KeyMaterial Keys { get; init; }
    public required string DeviceId { get; init; }
    public required string SaltB64 { get; init; }
    public required Func<AppSettings> Settings { get; init; }
}

public sealed class SyncEngine
{
    private readonly ApiClient _api;
    private readonly HistoryStore _store;
    private readonly ContentFetcher _fetcher;
    private readonly IClipboardSink _sink;
    private readonly IUserNotifier _notifier;
    private readonly EchoGuard _guard;
    private readonly IClock _clock;
    private readonly ILog _log;
    private readonly Func<SyncContext?> _context;
    private readonly object _gate = new();
    private readonly SemaphoreSlim _writeLock = new(1, 1);
    private readonly SemaphoreSlim _reconcileLock = new(1, 1);

    private SyncState _state;
    private bool _catchingUp;
    private readonly List<ItemHeader> _liveDuringCatchUp = new();
    private CancellationTokenSource? _welcomeCts;
    private Task _welcomeTask = Task.CompletedTask;
    private long _lastWrittenSeq;
    private TimeSpan _clockOffset;

    public SyncEngine(
        ApiClient api,
        HistoryStore store,
        ContentFetcher fetcher,
        IClipboardSink sink,
        IUserNotifier notifier,
        EchoGuard guard,
        IClock clock,
        ILog log,
        Func<SyncContext?> context)
    {
        _api = api;
        _store = store;
        _fetcher = fetcher;
        _sink = sink;
        _notifier = notifier;
        _guard = guard;
        _clock = clock;
        _log = log;
        _context = context;
        _state = store.LoadSyncState();
    }

    public int CatchUpPageLimit { get; set; } = ProtocolConstants.HistoryMaxLimit;

    public int MaxMissingFetch { get; set; } = 100;

    public Func<string, string>? DeviceNameResolver { get; set; }

    public TimeSpan ClockOffset
    {
        get
        {
            lock (_gate)
            {
                return _clockOffset;
            }
        }
    }

    public DateTimeOffset ServerNow => _clock.UtcNow + ClockOffset;

    public SyncState StateSnapshot
    {
        get
        {
            lock (_gate)
            {
                return new SyncState
                {
                    ServerId = _state.ServerId,
                    LastSeq = _state.LastSeq,
                    StateRev = _state.StateRev,
                    AppliedSeq = _state.AppliedSeq,
                };
            }
        }
    }

    public bool IsCatchingUp
    {
        get
        {
            lock (_gate)
            {
                return _catchingUp;
            }
        }
    }

    public Task WelcomeTask
    {
        get
        {
            lock (_gate)
            {
                return _welcomeTask;
            }
        }
    }

    public event Action<WelcomeMessage>? Welcomed;
    public event Action<PresenceMessage>? PresenceChanged;
    public event Action? DevicesChanged;
    public event Action<StorageWarningMessage>? StorageWarning;
    public event Action<string>? KeyInvalid;
    public event Action<HistoryEntry, ReceivedContent>? Applied;
    public event Action<bool>? SyncingChanged;
    public event Action<string>? ReleaseAvailable;

    public void ResetState()
    {
        lock (_gate)
        {
            _state = new SyncState();
            _store.SaveSyncState(_state);
            _lastWrittenSeq = 0;
        }
    }

    public void CancelPending()
    {
        CancellationTokenSource? cts;
        lock (_gate)
        {
            cts = _welcomeCts;
            _welcomeCts = null;
        }
        cts?.Cancel();
    }

    public void HandleMessage(string type, string json)
    {
        try
        {
            switch (type)
            {
                case WsTypes.Welcome:
                    {
                        var msg = JsonSerializer.Deserialize(json, ProtocolJsonContext.Default.WelcomeMessage);
                        if (msg != null)
                        {
                            OnWelcome(msg);
                        }
                        break;
                    }
                case WsTypes.Clip:
                    {
                        var msg = JsonSerializer.Deserialize(json, ProtocolJsonContext.Default.ClipMessage);
                        if (msg?.Item != null)
                        {
                            OnLiveClip(msg.Item);
                        }
                        break;
                    }
                case WsTypes.ClipDeleted:
                    {
                        var msg = JsonSerializer.Deserialize(json, ProtocolJsonContext.Default.ClipDeletedMessage);
                        if (msg != null)
                        {
                            _store.Delete(msg.Ids);
                            OnStateRevEvent(msg.StateRev);
                        }
                        break;
                    }
                case WsTypes.ClipPinned:
                    {
                        var msg = JsonSerializer.Deserialize(json, ProtocolJsonContext.Default.ClipPinnedMessage);
                        if (msg != null)
                        {
                            _store.SetPinned(msg.Id, msg.Pinned);
                            OnStateRevEvent(msg.StateRev);
                        }
                        break;
                    }
                case WsTypes.Presence:
                    {
                        var msg = JsonSerializer.Deserialize(json, ProtocolJsonContext.Default.PresenceMessage);
                        if (msg != null)
                        {
                            PresenceChanged?.Invoke(msg);
                        }
                        break;
                    }
                case WsTypes.DevicesChanged:
                    DevicesChanged?.Invoke();
                    break;
                case WsTypes.StorageWarning:
                    {
                        var msg = JsonSerializer.Deserialize(json, ProtocolJsonContext.Default.StorageWarningMessage);
                        if (msg != null)
                        {
                            StorageWarning?.Invoke(msg);
                        }
                        break;
                    }
                case WsTypes.ReleaseAvailable:
                    {
                        var msg = JsonSerializer.Deserialize(json, ProtocolJsonContext.Default.ReleaseAvailableMessage);
                        if (msg != null)
                        {
                            ReleaseAvailable?.Invoke(msg.Version);
                        }
                        break;
                    }
                case WsTypes.Error:
                    {
                        var msg = JsonSerializer.Deserialize(json, ProtocolJsonContext.Default.WsErrorMessage);
                        _log.Warn("sync", "server reported " + msg?.Code + ": " + msg?.Message);
                        break;
                    }
            }
        }
        catch (JsonException ex)
        {
            _log.Warn("sync", "could not parse " + type + " message", ex);
        }
        catch (Exception ex)
        {
            _log.Error("sync", "failed to handle " + type, ex);
        }
    }

    private void OnWelcome(WelcomeMessage welcome)
    {
        CancellationTokenSource cts;
        lock (_gate)
        {
            _welcomeCts?.Cancel();
            _welcomeCts = cts = new CancellationTokenSource();
            _catchingUp = true;
            _liveDuringCatchUp.Clear();
            _clockOffset = welcome.ServerTime - _clock.UtcNow;
        }
        _log.Info("sync", $"welcome: server seq {welcome.CurrentSeq}, state_rev {welcome.StateRev}, {welcome.OnlineDevices.Count} online");
        try
        {
            Welcomed?.Invoke(welcome);
        }
        catch
        {
        }
        var previous = WelcomeTask;
        var task = Task.Run(async () =>
        {
            try
            {
                await previous.ConfigureAwait(false);
            }
            catch
            {
            }
            await RunWelcomeAsync(welcome, cts.Token).ConfigureAwait(false);
        });
        lock (_gate)
        {
            _welcomeTask = task;
        }
    }

    private async Task RunWelcomeAsync(WelcomeMessage welcome, CancellationToken ct)
    {
        SyncingChanged?.Invoke(true);
        try
        {
            for (var attempt = 0; ; attempt++)
            {
                try
                {
                    await CatchUpAsync(welcome, ct).ConfigureAwait(false);
                    return;
                }
                catch (OperationCanceledException) when (ct.IsCancellationRequested)
                {
                    return;
                }
                catch (ApiException ex) when (ex.Status == 401)
                {
                    return;
                }
                catch (Exception ex) when (attempt < 6)
                {
                    var delay = RetryPolicy.HttpDelay(attempt);
                    _log.Warn("sync", $"catch-up failed, retrying in {delay.TotalSeconds:0.0}s: {ex.Message}");
                    try
                    {
                        await Task.Delay(delay, ct).ConfigureAwait(false);
                    }
                    catch (OperationCanceledException)
                    {
                        return;
                    }
                }
                catch (Exception ex)
                {
                    _log.Error("sync", "catch-up failed, waiting for the next connection", ex);
                    return;
                }
            }
        }
        finally
        {
            SyncingChanged?.Invoke(false);
        }
    }

    private async Task CatchUpAsync(WelcomeMessage welcome, CancellationToken ct)
    {
        var ctx = _context() ?? throw new InvalidOperationException("not signed in");
        SyncState state;
        lock (_gate)
        {
            state = _state;
        }
        if (state.ServerId != null && !string.Equals(state.ServerId, welcome.ServerId, StringComparison.Ordinal))
        {
            _log.Warn("sync", "server id changed, clearing local history");
            _store.Clear();
            lock (_gate)
            {
                _state = new SyncState { ServerId = null };
                _lastWrittenSeq = 0;
                _store.SaveSyncState(_state);
            }
            var me = await _api.GetMeAsync(ct).ConfigureAwait(false);
            if (!string.Equals(me.Salt, ctx.SaltB64, StringComparison.Ordinal) ||
                (me.KeyCheck != null && !string.Equals(me.KeyCheck, ctx.Keys.KeyCheck, StringComparison.Ordinal)))
            {
                _log.Warn("sync", "encryption salt or key check changed on the server");
                KeyInvalid?.Invoke("The server was reset. Enter the encryption password again.");
                return;
            }
            if (me.KeyCheck == null)
            {
                try
                {
                    await _api.PutKeyCheckAsync(ctx.Keys.KeyCheck, ct).ConfigureAwait(false);
                }
                catch (ApiException ex) when (ex.Code == ErrorCodes.KeyCheckExists)
                {
                    KeyInvalid?.Invoke("Another device set a different encryption password.");
                    return;
                }
            }
        }
        lock (_gate)
        {
            _state.ServerId = welcome.ServerId;
            _store.SaveSyncState(_state);
            state = _state;
        }
        var inserted = new List<ItemHeader>();
        var initial = state.LastSeq == 0;
        if (initial)
        {
            await InitialSyncAsync(welcome.CurrentSeq, ctx, ctx.Settings().LocalHistoryLimit, ct).ConfigureAwait(false);
            lock (_gate)
            {
                var liveMax = AutoApplyPolicy.HighestSeq(_liveDuringCatchUp, 0);
                _state.LastSeq = Math.Max(welcome.CurrentSeq, liveMax);
                _state.AppliedSeq = Math.Max(_state.AppliedSeq, welcome.CurrentSeq);
                _store.SaveSyncState(_state);
            }
        }
        else if (welcome.CurrentSeq > state.LastSeq)
        {
            var cursor = state.LastSeq;
            while (true)
            {
                ct.ThrowIfCancellationRequested();
                var page = await _api.GetHistoryAsync(after: cursor, limit: CatchUpPageLimit, ct: ct).ConfigureAwait(false);
                if (page.Items.Count == 0)
                {
                    break;
                }
                var entries = page.Items.Select(h => ItemCodec.Decrypt(h, ctx.Keys)).ToList();
                _store.UpsertMany(entries);
                foreach (var h in page.Items)
                {
                    _guard.Remember(h.ContentHash);
                }
                inserted.AddRange(page.Items);
                cursor = page.Items[^1].Seq;
                if (!page.HasMore)
                {
                    break;
                }
            }
            lock (_gate)
            {
                var liveMax = AutoApplyPolicy.HighestSeq(_liveDuringCatchUp, 0);
                _state.LastSeq = Math.Max(Math.Max(cursor, liveMax), _state.LastSeq);
                _store.SaveSyncState(_state);
            }
            _log.Info("sync", $"caught up {inserted.Count} items, last_seq {StateSnapshot.LastSeq}");
        }
        bool needReconcile;
        lock (_gate)
        {
            needReconcile = _state.StateRev == null || _state.StateRev.Value != welcome.StateRev;
        }
        if (needReconcile)
        {
            await ReconcileAsync(ct).ConfigureAwait(false);
        }
        ItemHeader? toApply;
        List<ItemHeader> candidates;
        lock (_gate)
        {
            ct.ThrowIfCancellationRequested();
            candidates = new List<ItemHeader>(inserted);
            candidates.AddRange(_liveDuringCatchUp);
            _liveDuringCatchUp.Clear();
            _catchingUp = false;
            toApply = AutoApplyPolicy.SelectAfterCatchUp(candidates, ctx.DeviceId, _state.AppliedSeq, _clock.UtcNow + _clockOffset);
            _state.AppliedSeq = AutoApplyPolicy.HighestSeq(candidates, _state.AppliedSeq);
            _state.LastSeq = AutoApplyPolicy.HighestSeq(candidates, _state.LastSeq);
            _store.SaveSyncState(_state);
        }
        if (toApply != null)
        {
            var entry = _store.Get(toApply.Id);
            if (entry != null)
            {
                _ = ApplyAsync(entry, toApply.Payload != null ? StrictBase64.Decode(toApply.Payload) : null, false, CancellationToken.None);
            }
        }
    }

    private async Task InitialSyncAsync(long currentSeq, SyncContext ctx, int maxItems, CancellationToken ct)
    {
        if (currentSeq <= 0)
        {
            return;
        }
        var before = currentSeq + 1;
        var total = 0;
        while (total < maxItems)
        {
            ct.ThrowIfCancellationRequested();
            var limit = Math.Min(CatchUpPageLimit, maxItems - total);
            var page = await _api.GetHistoryAsync(before: before, limit: limit, ct: ct).ConfigureAwait(false);
            if (page.Items.Count == 0)
            {
                break;
            }
            var entries = page.Items.Select(h => ItemCodec.Decrypt(h, ctx.Keys)).ToList();
            _store.UpsertMany(entries);
            total += page.Items.Count;
            before = page.Items.Min(h => h.Seq);
            if (!page.HasMore)
            {
                break;
            }
        }
        _log.Info("sync", $"initial sync loaded {total} items");
    }

    public async Task ReconcileAsync(CancellationToken ct)
    {
        var ctx = _context();
        if (ctx == null)
        {
            return;
        }
        await _reconcileLock.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            var index = await _api.GetHistoryIndexAsync(ct).ConfigureAwait(false);
            var local = _store.Index();
            var serverIds = new HashSet<string>(index.Items.Select(i => i.Id), StringComparer.Ordinal);
            var toDelete = local.Where(kv => !serverIds.Contains(kv.Key) && kv.Value.Seq <= index.CurrentSeq).Select(kv => kv.Key).ToList();
            if (toDelete.Count > 0)
            {
                _store.Delete(toDelete);
            }
            var pins = new Dictionary<string, bool>(StringComparer.Ordinal);
            foreach (var item in index.Items)
            {
                if (local.TryGetValue(item.Id, out var l) && l.Pinned != item.Pinned)
                {
                    pins[item.Id] = item.Pinned;
                }
            }
            if (pins.Count > 0)
            {
                _store.SetPinnedMany(pins);
            }
            var minSeq = local.Count > 0 ? local.Values.Min(v => v.Seq) : long.MaxValue;
            var missing = index.Items
                .Where(i => !local.ContainsKey(i.Id) && (i.Seq >= minSeq || i.Pinned))
                .OrderByDescending(i => i.Seq)
                .Take(MaxMissingFetch)
                .ToList();
            foreach (var m in missing)
            {
                try
                {
                    var header = await _api.GetItemAsync(m.Id, ct).ConfigureAwait(false);
                    var entry = ItemCodec.Decrypt(header, ctx.Keys);
                    _store.Upsert(entry, header.Payload != null && StrictBase64.TryDecode(header.Payload, out var p) ? p : null);
                }
                catch (ApiException ex) when (ex.Status == 404)
                {
                }
            }
            lock (_gate)
            {
                _state.StateRev = index.StateRev;
                _store.SaveSyncState(_state);
            }
            _log.Info("sync", $"reconciled: removed {toDelete.Count}, pin changes {pins.Count}, fetched {missing.Count}, state_rev {index.StateRev}");
        }
        finally
        {
            _reconcileLock.Release();
        }
    }

    private void OnStateRevEvent(long eventRev)
    {
        StateRevAction action;
        lock (_gate)
        {
            action = StateRevPolicy.OnEvent(_state.StateRev, eventRev);
            if (action == StateRevAction.Store)
            {
                _state.StateRev = eventRev;
                _store.SaveSyncState(_state);
            }
            if (_catchingUp && action == StateRevAction.Reconcile)
            {
                action = StateRevAction.Ignore;
            }
        }
        if (action == StateRevAction.Reconcile)
        {
            _log.Info("sync", "missed state changes, reconciling");
            _ = Task.Run(async () =>
            {
                try
                {
                    await ReconcileAsync(CancellationToken.None).ConfigureAwait(false);
                }
                catch (Exception ex)
                {
                    _log.Warn("sync", "reconcile failed", ex);
                }
            });
        }
    }

    private void OnLiveClip(ItemHeader header)
    {
        var ctx = _context();
        if (ctx == null)
        {
            return;
        }
        var entry = ItemCodec.Decrypt(header, ctx.Keys);
        byte[]? sealedPayload = null;
        if (header.Payload != null && !StrictBase64.TryDecode(header.Payload, out sealedPayload))
        {
            _log.Warn("sync", "invalid payload encoding for " + header.Id);
            sealedPayload = null;
        }
        _store.Upsert(entry, sealedPayload);
        _guard.Remember(header.ContentHash);
        bool eligible;
        lock (_gate)
        {
            if (_catchingUp)
            {
                _liveDuringCatchUp.Add(header);
                return;
            }
            _state.LastSeq = Math.Max(_state.LastSeq, header.Seq);
            eligible = AutoApplyPolicy.IsEligible(header, ctx.DeviceId, _state.AppliedSeq, _clock.UtcNow + _clockOffset);
            _state.AppliedSeq = Math.Max(_state.AppliedSeq, header.Seq);
            _store.SaveSyncState(_state);
        }
        if (eligible)
        {
            _ = ApplyAsync(entry, sealedPayload, false, CancellationToken.None);
        }
    }

    public async Task<bool> ApplyAsync(HistoryEntry entry, byte[]? sealedPayload, bool manual, CancellationToken ct)
    {
        var ctx = _context();
        if (ctx == null || entry.Meta == null)
        {
            return false;
        }
        var settings = ctx.Settings();
        if (!manual)
        {
            if (!settings.AutoApply)
            {
                return false;
            }
            var kindAllowed = entry.Header.Kind switch
            {
                ProtocolConstants.KindText => settings.SyncText,
                ProtocolConstants.KindImage => settings.SyncImages,
                ProtocolConstants.KindFiles => settings.SyncFiles,
                _ => false,
            };
            if (!kindAllowed)
            {
                _log.Info("apply", "skipping " + entry.Header.Kind + " item, receiving this kind is off");
                return false;
            }
            if (entry.Header.ChunkCount > 0 && entry.Header.Size > settings.AutoDownloadLimitBytes)
            {
                _log.Info("apply", $"item {entry.Id} is {HistoryEntry.FormatSize(entry.Header.Size)}, above the auto-download limit");
                _notifier.LargeItemAvailable(entry, DeviceNameResolver?.Invoke(entry.Header.DeviceId) ?? "another device");
                return false;
            }
        }
        try
        {
            using var fetched = await _fetcher.FetchAsync(entry, ctx.Keys, sealedPayload, null, ct).ConfigureAwait(false);
            await _writeLock.WaitAsync(ct).ConfigureAwait(false);
            try
            {
                if (!manual && entry.Seq < _lastWrittenSeq)
                {
                    _log.Info("apply", $"skipping {entry.Id}, a newer item was applied meanwhile");
                    return false;
                }
                var content = _fetcher.Materialize(entry, fetched);
                _guard.Remember(entry.Header.ContentHash);
                await _sink.WriteAsync(content, ct).ConfigureAwait(false);
                _lastWrittenSeq = Math.Max(_lastWrittenSeq, entry.Seq);
                _log.Info("apply", $"placed {entry.KindLabel.ToLowerInvariant()} item seq {entry.Seq} on the clipboard");
                try
                {
                    Applied?.Invoke(entry, content);
                }
                catch
                {
                }
                return true;
            }
            finally
            {
                _writeLock.Release();
            }
        }
        catch (CryptoException ex)
        {
            _log.Error("apply", "item " + entry.Id + " failed verification, discarded", ex);
            return false;
        }
        catch (OperationCanceledException)
        {
            return false;
        }
        catch (Exception ex)
        {
            _log.Error("apply", "could not apply item " + entry.Id, ex);
            if (manual)
            {
                throw;
            }
            return false;
        }
    }
}
